src/routing/bootstrap.rs

```diff
@@ -476,72 +368,77 @@ impl<'a> State<'a> {
         expected_genesis_key: Option<&bls::PublicKey>,
         relocate_payload: Option<&RelocatePayload>,
     ) -> Result<(JoinResponse, SocketAddr, DestInfo)> {
+        let destination = match relocate_payload {
+            Some(payload) => *payload.details.destination()?,
+            None => self.node.name(),
+        };
+
         while let Some((message, sender)) = self.recv_rx.next().await {
-            let (message, dest_info) = match message {
-                MessageType::Routing { msg, dest_info } => (msg, dest_info),
+            // we are interested only in `JoinResponse` type of messages
+            let (routing_msg, dest_info, join_response) = match message {
                 MessageType::Node { .. }
                 | MessageType::Client { .. }
                 | MessageType::SectionInfo { .. } => continue,
+                MessageType::Routing { msg, dest_info } => {
+                    if let Variant::JoinResponse(resp) = &msg.variant {
+                        let join_response = resp.clone();
+                        (msg, dest_info, *join_response)
+                    } else {
+                        self.backlog_message(msg, sender, dest_info);
+                        continue;
+                    }
+                }
             };
 
-            match message.variant() {
-                Variant::JoinRetry {
-                    section_auth,
-                    section_key,
-                } => {
-                    if !self.verify_message(&message, None) {
+            match join_response {
+                JoinResponse::Rejected(JoinRejectionReason::NodeNotReachable(_))
+                | JoinResponse::Rejected(JoinRejectionReason::JoinsDisallowed) => {
+                    return Ok((join_response, sender, dest_info));
```

Comment:
the above `used_recipients` filter shall only be carried out for the case of `NodeNotReachable`